onnxruntime/core/graph/graph.cc

```diff
@@ -4044,13 +4043,25 @@ Status Graph::InlineFunction(Node& callnode) {
       return &this->GetOrCreateNodeArg(name, nullptr);
     };
 
+    // Process constant nodes first and create NodeArg for these as they become initializers
+    // It is important for the initializers to have NodeArg created, first they are needed
+    // if the initializer is unused and removed, second if the node depends on the initializer,
+    // we can have Type attached to it.
     for (const auto& inlined_node : inlined_fp.node()) {
       if (inlined_node.op_type() == kConstant) {
         // Copy constant nodes _value to name_to_initial_tensor_
         const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
         ORT_RETURN_IF_ERROR(utils::ConstantNodeProtoToTensorProto(inlined_node, model_path, *tensor, inlined_node.output(0)));
-        name_to_initial_tensor_[tensor->name()] = tensor;
-      } else {
+        auto insert_result = name_to_initial_tensor_.emplace(tensor->name(), tensor);
+        ORT_RETURN_IF_NOT(insert_result.second, "Constant node name: ", tensor->name(), " in inlined function: ",
```

Comment:
Not sure If I understand code and context correctly. It looks like now we return failure if there is duplicate constant node name, whereas previous code overrides the initializer.

How does this work with the following two scenarios?
- Models with different function protos with inner constants with same names. How to distinguish between them?
- Models with multiple function nodes of the same function proto. Will it try to inline same function proto multiple times and trigger a failure when name conflicts?

Reply:
These are good questions. One thing is the function local initializer cannot simply override the global initializer. I am seeing that the names are currently very elaborate and scoped, but that would not be enough to inline a function multiple times.
Would be great to see a good document on Functions, that describes all those cases.

Reply:
I think constant values remain the same for the same function_proto. However, the spec does not and cannot ask for constant names within function_proto to be unique across the entire model. The constant names are only unique within the immediate scope of that function itself. @gramalingam please chime in on how this is formally defined in function spec.

Reply:
Can we use a combination of function proto's name, domain and constant node's output name for the new initializer name? This should be unique across different functions.

Reply:
We need to account for multiple inlining of the same function. This will add complexity. My understanding is that we would need to replace references to the constant nodes and initializers within the function.

Reply:
Names are made unique here: https://github.com/microsoft/onnxruntime/blob/yuslepukhin/inlined_initializers_nodearg/onnxruntime/core/graph/function_utils.cc

Reply:
If `function_utils::Specialize` has applied the unique identifier (`auto uniq_identifier = GenerateNodeName(ss.str());` all the values being inlined will already (probably) be a unique name won't they? 

[side note: `function_utils::Specialize` would need the Graph instance to be able to validate any generated names are definitely unique by calling GetNodeArg and making sure that returns nullptr]

So in theory shouldn't clash with an existing value but inefficient when the same function gets inlined multiple times.

Could maybe do something like Pranav suggested but may need to use the node's index in the function proto instead of the output value (as I think the value as been uniqified by then) to consistently identify initializer across multiple usages of the function proto. But if you did that how do you know whether the generated value actually existed in the graph to start with (unlikely but in theory...) vs being from the previous inlining of the same function? 

Is this a problem we have to solve? Naively the initializers in a function that is inlined multiple times are going to be small as they will be simple things like values for axis/dims/indexes/settings rather than weights. 

I don't think we can error out if we find a conflict. We need to avoid the conflict when inlining instead.

Reply:
Yes, I agree. Some checks can be performed after Specialize has finished. We need to avoid the conflict, but verify still.
Not all of the initializers will be small. The model I am debugging now has lots of `If` nodes each with 2 `Constant` nodes, with what it looks like transposed dimensions. It is either [768, 768], [768, 3072] or [768, 50265]

Reply:
Yes, names must be made unique when inlining happens. Originally, there was a heuristic that used pointer-addresses for creating unique-names during inlining (which is not guaranteed to work always). A few months back there was a PR that replaced this with another renaming strategy. I haven't had a chance to look at that renaming strategy properly, but I did have some concerns about it back then (I was on vacation when the PR was merged). But this issue is fairly restricted in scope, it is just a question of how to generate unique-names, which is solvable in several ways.

Reply:
There is also an issue of creating NodeArgs for everything here, and potentially some other structures that won't be needed after inlining. For example, lots of initializers are removed as un-used which exposed the fact that NodeArgs were not created for them.

Reply:
Just to clarify: this PR is fine, the issue being discussed is a pre-existing one (in the name-uniquification strategy, which should work in practice most of the time). The above check is just validating that the name-uniquification did a correct job (so, is really an ORT_ENFORCE).